test/common/network/udp_listener_impl_test.cc

```diff
@@ -505,6 +423,73 @@ TEST_P(UdpListenerImplTest, UdpListenerRecvFromError) {
   dispatcher_->run(Event::Dispatcher::RunType::Block);
 }
 
+/**
+ * Tests UDP listener for sending datagrams to destination.
+ *  1. Setup a server udp listener and client udp listener
+ *  2. Send the data from the client component to the server and validate the contents
+ */
+TEST_P(UdpListenerImplTest, SendData) {
+  // Setup server socket
+  SocketPtr server_socket =
+      getSocket(Address::SocketType::Datagram, Network::Test::getCanonicalLoopbackAddress(version_),
+                nullptr, true);
+
+  ASSERT_NE(server_socket, nullptr);
+
+  auto const* server_ip = server_socket->localAddress()->ip();
+  ASSERT_NE(server_ip, nullptr);
+
+  // Setup server callback handler and listener.
+  Network::MockUdpListenerCallbacks server_listener_callbacks;
+  Network::TestUdpListenerImpl server_listener(dispatcherImpl(), *server_socket.get(),
+                                               server_listener_callbacks);
+
+  EXPECT_CALL(server_listener, doRecvFrom(_, _))
+      .WillRepeatedly(Invoke([&](sockaddr_storage& peer_addr, socklen_t& addr_len) {
+        return server_listener.doRecvFrom_(peer_addr, addr_len);
+      }));
+
+  // Setup client socket.
+  SocketPtr client_socket =
+      getSocket(Address::SocketType::Datagram, Network::Test::getCanonicalLoopbackAddress(version_),
+                nullptr, false);
+
+  ASSERT_NE(client_socket, nullptr);
+
+  // Setup client callback handler and listener.
+  Network::MockUdpListenerCallbacks client_listener_callbacks;
+  Network::TestUdpListenerImpl client_listener(dispatcherImpl(), *client_socket.get(),
+                                               client_listener_callbacks);
+
+  const std::string payload("hello world");
+  Buffer::InstancePtr buffer(new Buffer::OwnedImpl());
+  buffer->add(payload);
+  UdpSendData send_data{server_socket->localAddress(), *buffer};
+
+  client_listener.send(send_data);
+
+  EXPECT_CALL(client_listener_callbacks, onError_(_, _)).Times(0);
+  EXPECT_CALL(server_listener_callbacks, onError_(_, _)).Times(0);
+
+  EXPECT_CALL(server_listener_callbacks, onWriteReady_(_))
```

Comment:
I thought onWriteReady() would only be trigger when socket gets unblocked which isn't the case for this test. Can you comment why does this test expect it on both server and client listeners.

Reply:
My understanding is that the base case when you register in epoll for a write/read event, the write event always fires because the FD is open for writes. Hence, the expect_call in the test.

The event is fired again only when a write fails with ETRYAGAIN.

Reply:
Thanks for explanation. Can you add a comment that the onWriteReady() is fired when it gets registered?

Reply:
sure.